Cython/Utility/TypeConversion.c

```diff
@@ -286,20 +291,61 @@ static CYTHON_INLINE const char* __Pyx_PyObject_AsStringAndSize(PyObject* o, Py_
         return PyByteArray_AS_STRING(o);
 #else
         *length = PyByteArray_Size(o);
-        if (*length == -1) return NULL;
+        if (*length == -1) goto error;
         return PyByteArray_AsString(o);
 #endif
 
-    } else
-    {
+    } else if (PyBytes_Check(o)) {
         char* result;
         int r = PyBytes_AsStringAndSize(o, &result, length);
         if (unlikely(r < 0)) {
-            return NULL;
+            goto error;
         } else {
             return result;
         }
+    } else {
+        char* result;
+        int r;
+#if !CYTHON_COMPILING_IN_LIMITED_API || __PYX_LIMITED_VERSION_HEX >= 0x030B0000
+        // Check that `o`:
+        // * Supports the Python Buffer Protocol
+        // * There is no need to keep the buffer around (as we cannot return it)
+        PyBufferProcs *pb = Py_TYPE(o)->tp_as_buffer;
+        if (pb == NULL || pb->bf_getbuffer == NULL || pb->bf_releasebuffer != NULL) {
+            goto error;
+        }
```

Comment:
Have added this guard to only use the Python Buffer Protocol in either non-Limited API builds or Limited API builds on Python 3.11+. In those cases this path should be ok

Also added an explicit check that the Python Buffer Protocol does work on this object. This check can only be done under the conditions laid out previously. However this does help us pick out any cases where the object doesn't support the Python Buffer Protocol or it does, but depends on releasing the buffer after

Please let me know if there are other cases we should check for or if this needs tweaking

Reply:
If you want this to work in the Limited API, you need to use `__Pyx_PyType_GetSlot()`.